### PR TITLE
sw_engine raster: ++safety of the scaled image rasterization

### PR DESCRIPTION
**Pull request by hermet**
Prevent out of buffer boundary access.

@Issues: https://github.com/Samsung/thorvg/issues/1118

### CODE REVIEW

**Comment by hermet** [comment]
Need additional patch to fix the #1118

File: src/lib/sw_engine/tvgSwRaster.cpp [hunk 1]
```diff
@@ -740,9 +748,11 @@ static bool _rasterScaledTranslucentRGBAImage(SwSurface* surface, const SwImage*
     if (image->scale < DOWN_SCALE_TOLERANCE) {
         for (auto y = region.min.y; y < region.max.y; ++y, dbuffer += surface->stride) {
             auto sy = static_cast<uint32_t>(y * itransform->e22 + itransform->e23);
+            if (sy >= image->h) continue;
             auto dst = dbuffer;
             for (auto x = region.min.x; x < region.max.x; ++x, ++dst) {
                 auto sx = static_cast<uint32_t>(x * itransform->e11 + itransform->e13);
```
Comment (JSUYA) [inline src/lib/sw_engine/tvgSwRaster.cpp:754]: I think the problem occurs when the result of (x * itransform->e11 + itransform->e13) is negative.
(or memory problem ?)
Wouldn't the result of this expression be checked first(0 > result || w < result ?)  and then cast?

**Review by JSUYA — COMMENTED** [review]
please check my comment

File: src/lib/sw_engine/tvgSwRaster.cpp (hunk 1 above)
Comment (hermet) [reply src/lib/sw_engine/tvgSwRaster.cpp:754]: we use "unsigned" so negative is not happening.

Comment (JSUYA) [reply src/lib/sw_engine/tvgSwRaster.cpp:754]: As far as I know there is a problem casting float to "unsigned" int.
please refer to https://embeddeduse.com/2013/08/25/casting-a-negative-float-to-an-unsigned-int/

Comment (hermet) [reply src/lib/sw_engine/tvgSwRaster.cpp]: @JSUYA  updated thanks.

File: src/lib/sw_engine/tvgSwRaster.cpp [hunk 2]
```diff
@@ -392,27 +395,27 @@ static bool _rasterScaledTranslucentRleRGBAImage(SwSurface* surface, const SwIma
     //Center (Down-Scaled)
     if (image->scale < DOWN_SCALE_TOLERANCE) {
         for (uint32_t i = 0; i < image->rle->size; ++i, ++span) {
-            auto sy = static_cast<uint32_t>(span->y * itransform->e22 + itransform->e23);
+            auto sy = (uint32_t)(span->y * itransform->e22 + itransform->e23);
```
Comment (JSUYA) [inline src/lib/sw_engine/tvgSwRaster.cpp:398]: Wouldn't the logic of this function be changed like other functions?

File: src/lib/sw_engine/tvgSwRaster.cpp [hunk 3]
```diff
@@ -429,42 +432,42 @@ static bool _rasterScaledRleRGBAImage(SwSurface* surface, const SwImage* image,
...
```
Comment (JSUYA) [inline src/lib/sw_engine/tvgSwRaster.cpp:435]: this function too.

File: src/lib/sw_engine/tvgSwRaster.cpp [hunk 4]
```diff
@@ -660,13 +663,15 @@ static bool _rasterScaledMaskedTranslucentRGBAImage(SwSurface* surface, const Sw
     // Down-Scaled
     if (image->scale < DOWN_SCALE_TOLERANCE) {
         for (auto y = region.min.y; y < region.max.y; ++y) {
+            auto sy = (uint32_t)(y * itransform->e22 + itransform->e23);
```
Comment (JSUYA) [inline src/lib/sw_engine/tvgSwRaster.cpp:666]: this function too

**Review by JSUYA — COMMENTED** [review]
Looks like some functions haven't been fixed yet. could you please check again?

**Review by JSUYA — APPROVED** [review]

**Pull request merged**